hera_cal/lst_stack/metrics.py

```diff
@@ -0,0 +1,386 @@
+        if std == 'autos':
+            variance = get_nightly_predicted_variance(bl, stacks, auto_stats, dt, df)
+        elif std in {'std', 'mad'}:
+            variance = np.abs(np.where(cross_stats.flags[bl], np.inf, getattr(cross_stats, std)[bl]))**2
+        else:
+            raise ValueError("std must be 'autos', 'std' or 'mad'")
+
+        for i, (zstack, v) in enumerate(zip(zstacks, variance)):
+            data = zstack.get_data(bl)
+            flg = zstack.get_flags(bl)
+
+            centre = getattr(cross_stats, central)[bl][i]
+
+            z = np.abs(data - centre)**2 / (v / 2)  # TODO: check that we need /2 for std, mad
+
+            z[flg] = np.nan
+
+            zstack.set_data(z[:, :, None], bl)
+
+    return zstacks
+
+
+def get_selected_bls(
+    bls: list[tuple[int, int, str]],
+    days_binned: DataContainer | None = None,
+    selectors=None,
+    min_days: int = 7
+):
+    if selectors is None:
+        selectors = []
+    elif callable(selectors):
+        selectors = [selectors]
+
+    if min_days:
+        selectors.append(lambda bl: bl in days_binned and (np.median(days_binned[bl]) >= min_days))
+
+    def select(bl):
+        return all(sel(bl) for sel in selectors)
+
+    return [bl for bl in bls if select(bl)]
+
+
+def _get_data_subset(
+    zscores: list[UVData],
+    bls: list[tuple[int, int, str]] | None = None,
+    band: tuple[int, int] | None = None,
+    nights='all',
+    pols='all',
+    selector=None
+):
+
+    allz = []
+
+    nbls = zscores[0].Nbls
+    datapols = utils.polnum2str(
+        zscores[0].polarization_array, x_orientation=zscores[0].x_orientation
+    )
+    datapairs = list(
+        zip(zscores[0].ant_1_array[:nbls], zscores[0].ant_2_array[:nbls])
+    )
+
+    if selector is not None:
+        allbls = [(a, b, p) for a, b in datapairs for p in datapols]
+        bls = get_selected_bls(allbls, min_days=0, selectors=selector)
+        selpols = {p for a, b, p in bls}
+        bls = list({bl[:2] for bl in bls})  # only antpairs
+
+        if 'ee' in selpols and 'nn' in selpols:
+            pols = 'all'
+        elif 'ee' in selpols:
+            pols = 'ee'
+        elif 'nn' in selpols:
+            pols = 'nn'
+
+    # Get pol indices
+    if pols == 'all':
+        pols = slice(None)
+    elif isinstance(pols, str):
+        pols = [datapols.index(pols)]
+    else:
+        pols = [datapols.index(p) for p in pols]
+
+    # Get bl indices
+    if bls is None:
+        bls = datapairs
+    if isinstance(bls, tuple) and len(bls) == 2:
+        bls = [datapairs.index(bls)]
+    elif isinstance(bls, list):
+        bls = [datapairs.index(bl) for bl in bls]
+
+    if band is None:
+        band = slice(None)
+    elif isinstance(band, tuple):
+        band = slice(band[0], band[1])
+
+    for lstbin, zscore in enumerate(zscores):
+        newshape = (zscore.Ntimes, zscore.Nbls, -1, zscore.Npols)
+        data = zscore.data_array[:, band].real.reshape(newshape)
```

Comment:
OK, so I've gone and renamed things better to emphasise we're working with zscores here. Really, we should be able to work with any `LSTStack` with the `.metrics` attribute defined, but this is exclusive going to be z^2 in this module/notebook, so probably ok to be less general.

This goes from UVData to array because I didn't want to have to properly select all the metadata etc -- the point of this function is just to return some zscores within the selection criteria, so that they can be plotted etc. You're not meant to have to know how to properly index the resulting array -- you're supposed to just use the numbers in it.